File: crossword.py
"""クロスワードパズル生成ロジック"""

import random

# 小さい仮名を通常サイズに変換（日本語クロスワードの慣例）
SMALL_TO_LARGE = str.maketrans("ぁぃぅぇぉっゃゅょ", "あいうえおつやゆよ")


def to_grid_form(word):
    """単語をクロスワード用の表記に変換する（小文字仮名→大文字仮名）"""
    return word.translate(SMALL_TO_LARGE)


ACROSS = "across"
DOWN = "down"


class CrosswordGenerator:
    def __init__(self, width, height, word_list):
        self.width = width
        self.height = height
        max_len = max(width, height)
        self.word_list = [
            (w, c) for w, c in word_list if len(to_grid_form(w)) <= max_len
        ]
        self.grid = None
        self.placed = None

    def generate(self, attempts=80):
        """複数回試行して最良の結果を返す"""
        best = None
        best_count = 0

        for _ in range(attempts):
            self._reset()
            self._try_generate()
            count = len(self.placed)
            if count > best_count:
                best_count = count
                best = (self._copy_grid(), list(self.placed))
            if count >= 12:
                break

        if best:
            self.grid, self.placed = best
            return self._build_result()

        return {"error": "パズルを生成できませんでした。マスのサイズを大きくしてください。"}

    def _reset(self):
        self.grid = [[None] * self.width for _ in range(self.height)]
        self.placed = []

    def _copy_grid(self):
        return [row[:] for row in self.grid]

    def _try_generate(self):
        words = list(self.word_list)
        random.shuffle(words)
        words.sort(key=lambda x: len(x[0]), reverse=True)

        if not words:
            return

        # 最初の単語を中央に横向きで配置
        first_word, first_clue = words[0]
        grid_word = to_grid_form(first_word)
        col = (self.width - len(grid_word)) // 2
        row = self.height // 2

        if col >= 0 and col + len(grid_word) <= self.width:
            self._place(first_word, first_clue, row, col, ACROSS)

        # 残りの単語を配置
        for word, clue in words[1:]:
            placements = self._find_placements(word)
            if placements:
                placements.sort(key=lambda p: p["score"], reverse=True)
                top = placements[: min(3, len(placements))]
                chosen = random.choice(top)
                self._place(word, clue, chosen["row"], chosen["col"], chosen["dir"])

    def _place(self, word, clue, row, col, direction):
        grid_word = to_grid_form(word)
        dr, dc = (0, 1) if direction == ACROSS else (1, 0)
        for i, ch in enumerate(grid_word):
            self.grid[row + dr * i][col + dc * i] = ch
        self.placed.append((word, clue, row, col, direction))

    def _find_placements(self, word):
        grid_word = to_grid_form(word)
        results = []

        for direction in [ACROSS, DOWN]:
            dr, dc = (0, 1) if direction == ACROSS else (1, 0)

            if direction == ACROSS:
                max_row, max_col = self.height - 1, self.width - len(grid_word)
            else:
                max_row, max_col = self.height - len(grid_word), self.width - 1

            if max_col < 0 or max_row < 0:
                continue

            for row in range(max_row + 1):
                for col in range(max_col + 1):
                    score = self._check_placement(grid_word, row, col, dr, dc)
                    if score is not None:
                        results.append(
                            {"row": row, "col": col, "dir": direction, "score": score}
                        )

        return results

    def _check_placement(self, grid_word, row, col, dr, dc):
        length = len(grid_word)

        # 単語の前のマスが空/端であること
        pr, pc = row - dr, col - dc
        if 0 <= pr < self.height and 0 <= pc < self.width:
            if self.grid[pr][pc] is not None:
                return None

        # 単語の後のマスが空/端であること
        ar, ac = row + dr * length, col + dc * length
        if 0 <= ar < self.height and 0 <= ac < self.width:
            if self.grid[ar][ac] is not None:
                return None

        intersections = 0
        for i, ch in enumerate(grid_word):
            r = row + dr * i
            c = col + dc * i

            cell = self.grid[r][c]
            if cell is not None:
                if cell == ch:
                    intersections += 1
                else:
                    return None
            else:
                # 垂直方向の隣接マスが空であること（意図しない単語の生成を防ぐ）
                if dr == 0:  # 横方向配置 → 上下チェック
                    if r > 0 and self.grid[r - 1][c] is not None:
                        return None
                    if r < self.height - 1 and self.grid[r + 1][c] is not None:
                        return None
                else:  # 縦方向配置 → 左右チェック
                    if c > 0 and self.grid[r][c - 1] is not None:
                        return None
                    if c < self.width - 1 and self.grid[r][c + 1] is not None:
                        return None

        # 最初の単語以外は必ず交差が必要
        if intersections == 0 and len(self.placed) > 0:
            return None

        return intersections

    def _build_result(self):
        # グリッドデータの構築
        grid = []
        for row in self.grid:
            grid.append([ch if ch is not None else "" for ch in row])

        # 配置単語の位置マップ
        word_map = {}
        for word, clue, row, col, direction in self.placed:
            word_map[(row, col, direction)] = (word, clue)

        # セル番号の割り当て
        cell_numbers = [[0] * self.width for _ in range(self.height)]
        clues_across = []
        clues_down = []
        answers = {}
        number = 1

        for r in range(self.height):
            for c in range(self.width):
                if self.grid[r][c] is None:
                    continue

                starts_across = False
                starts_down = False

                # 横の単語の開始判定
                if (c == 0 or self.grid[r][c - 1] is None) and (
                    c + 1 < self.width and self.grid[r][c + 1] is not None
                ):
                    starts_across = True

                # 縦の単語の開始判定
                if (r == 0 or self.grid[r - 1][c] is None) and (
                    r + 1 < self.height and self.grid[r + 1][c] is not None
                ):
                    starts_down = True

                if starts_across or starts_down:
                    cell_numbers[r][c] = number

                    if starts_across and (r, c, ACROSS) in word_map:
                        word, clue = word_map[(r, c, ACROSS)]
                        grid_word = to_grid_form(word)
                        clues_across.append({"number": number, "clue": clue})
                        answers[f"{number}-across"] = grid_word

                    if starts_down and (r, c, DOWN) in word_map:
                        word, clue = word_map[(r, c, DOWN)]
                        grid_word = to_grid_form(word)
                        clues_down.append({"number": number, "clue": clue})
                        answers[f"{number}-down"] = grid_word

                    number += 1

        return {
            "width": self.width,
            "height": self.height,
            "grid": grid,
            "cell_numbers": cell_numbers,
            "clues_across": clues_across,
            "clues_down": clues_down,
            "answers": answers,
        }

File: static/crossword.js
// クロスワードパズル生成ロジック

const SMALL_TO_LARGE = {
  "ぁ": "あ", "ぃ": "い", "ぅ": "う", "ぇ": "え", "ぉ": "お",
  "っ": "つ", "ゃ": "や", "ゅ": "ゆ", "ょ": "よ",
};

function toGridForm(word) {
  let out = "";
  for (const ch of word) {
    out += SMALL_TO_LARGE[ch] || ch;
  }
  return out;
}

const ACROSS = "across";
const DOWN = "down";

function shuffled(arr) {
  const a = arr.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

class CrosswordGenerator {
  constructor(width, height, wordList) {
    this.width = width;
    this.height = height;
    const maxLen = Math.max(width, height);
    this.wordList = wordList.filter(([w]) => toGridForm(w).length <= maxLen);
    this.grid = null;
    this.placed = null;
  }

  generate(attempts = 80) {
    let best = null;
    let bestCount = 0;

    for (let i = 0; i < attempts; i++) {
      this._reset();
      this._tryGenerate();
      const count = this.placed.length;
      if (count > bestCount) {
        bestCount = count;
        best = [this._copyGrid(), this.placed.slice()];
      }
      if (count >= 12) break;
    }

    if (best) {
      this.grid = best[0];
      this.placed = best[1];
      return this._buildResult();
    }

    return { error: "パズルを生成できませんでした。マスのサイズを大きくしてください。" };
  }

  _reset() {
    this.grid = Array.from({ length: this.height }, () =>
      new Array(this.width).fill(null)
    );
    this.placed = [];
  }

  _copyGrid() {
    return this.grid.map((row) => row.slice());
  }

  _tryGenerate() {
    let words = shuffled(this.wordList);
    if (words.length === 0) return;

    const firstCandidates = words.filter(
      ([w]) => toGridForm(w).length <= this.width
    );
    if (firstCandidates.length === 0) return;

    const [firstWord, firstClue] =
      firstCandidates[Math.floor(Math.random() * firstCandidates.length)];
    words = words.filter(([w]) => w !== firstWord);

    const gridWord = toGridForm(firstWord);
    const col = Math.floor((this.width - gridWord.length) / 2);
    const row = Math.floor(this.height / 2);
    this._place(firstWord, firstClue, row, col, ACROSS);

    for (const [word, clue] of words) {
      const placements = this._findPlacements(word);
      if (placements.length > 0) {
        placements.sort((a, b) => b.score - a.score);
        const top = placements.slice(0, Math.min(3, placements.length));
        const chosen = top[Math.floor(Math.random() * top.length)];
        this._place(word, clue, chosen.row, chosen.col, chosen.dir);
      }
    }
  }

  _place(word, clue, row, col, direction) {
    const gridWord = toGridForm(word);
    const dr = direction === ACROSS ? 0 : 1;
    const dc = direction === ACROSS ? 1 : 0;
    for (let i = 0; i < gridWord.length; i++) {
      this.grid[row + dr * i][col + dc * i] = gridWord[i];
    }
    this.placed.push([word, clue, row, col, direction]);
  }

  _findPlacements(word) {
    const gridWord = toGridForm(word);
    const results = [];

    for (const direction of [ACROSS, DOWN]) {
      const dr = direction === ACROSS ? 0 : 1;
      const dc = direction === ACROSS ? 1 : 0;

      let maxRow, maxCol;
      if (direction === ACROSS) {
        maxRow = this.height - 1;
        maxCol = this.width - gridWord.length;
      } else {
        maxRow = this.height - gridWord.length;
        maxCol = this.width - 1;
      }

      if (maxCol < 0 || maxRow < 0) continue;

      for (let row = 0; row <= maxRow; row++) {
        for (let col = 0; col <= maxCol; col++) {
          const score = this._checkPlacement(gridWord, row, col, dr, dc);
          if (score !== null) {
            results.push({ row, col, dir: direction, score });
          }
        }
      }
    }

    return results;
  }

  _checkPlacement(gridWord, row, col, dr, dc) {
    const length = gridWord.length;

    const pr = row - dr;
    const pc = col - dc;
    if (pr >= 0 && pr < this.height && pc >= 0 && pc < this.width) {
      if (this.grid[pr][pc] !== null) return null;
    }

    const ar = row + dr * length;
    const ac = col + dc * length;
    if (ar >= 0 && ar < this.height && ac >= 0 && ac < this.width) {
      if (this.grid[ar][ac] !== null) return null;
    }

    let intersections = 0;
    for (let i = 0; i < length; i++) {
      const r = row + dr * i;
      const c = col + dc * i;
      const cell = this.grid[r][c];
      const ch = gridWord[i];

      if (cell !== null) {
        if (cell === ch) {
          intersections++;
        } else {
          return null;
        }
      } else {
        if (dr === 0) {
          if (r > 0 && this.grid[r - 1][c] !== null) return null;
          if (r < this.height - 1 && this.grid[r + 1][c] !== null) return null;
        } else {
          if (c > 0 && this.grid[r][c - 1] !== null) return null;
          if (c < this.width - 1 && this.grid[r][c + 1] !== null) return null;
        }
      }
    }

    if (intersections === 0 && this.placed.length > 0) return null;

    return intersections;
  }

  _buildResult() {
    const grid = this.grid.map((row) =>
      row.map((ch) => (ch !== null ? ch : ""))
    );

    const wordMap = new Map();
    for (const [word, clue, row, col, direction] of this.placed) {
      wordMap.set(`${row},${col},${direction}`, [word, clue]);
    }

    const cellNumbers = Array.from({ length: this.height }, () =>
      new Array(this.width).fill(0)
    );
    const cluesAcross = [];
    const cluesDown = [];
    const answers = {};
    let number = 1;

    for (let r = 0; r < this.height; r++) {
      for (let c = 0; c < this.width; c++) {
        if (this.grid[r][c] === null) continue;

        const startsAcross =
          (c === 0 || this.grid[r][c - 1] === null) &&
          c + 1 < this.width &&
          this.grid[r][c + 1] !== null;

        const startsDown =
          (r === 0 || this.grid[r - 1][c] === null) &&
          r + 1 < this.height &&
          this.grid[r + 1][c] !== null;

        if (startsAcross || startsDown) {
          cellNumbers[r][c] = number;

          if (startsAcross) {
            const key = `${r},${c},${ACROSS}`;
            if (wordMap.has(key)) {
              const [word, clue] = wordMap.get(key);
              cluesAcross.push({ number, clue });
              answers[`${number}-across`] = toGridForm(word);
            }
          }

          if (startsDown) {
            const key = `${r},${c},${DOWN}`;
            if (wordMap.has(key)) {
              const [word, clue] = wordMap.get(key);
              cluesDown.push({ number, clue });
              answers[`${number}-down`] = toGridForm(word);
            }
          }

          number++;
        }
      }
    }

    return {
      width: this.width,
      height: this.height,
      grid,
      cell_numbers: cellNumbers,
      clues_across: cluesAcross,
      clues_down: cluesDown,
      answers,
    };
  }
}

function generateCrossword(width, height) {
  width = Math.max(5, Math.min(15, width));
  height = Math.max(5, Math.min(15, height));
  const generator = new CrosswordGenerator(width, height, WORDS);
  return generator.generate();
}
